contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -4453,3 +4608,691 @@ rename_tsql_db(char *old_db_name, char *new_db_name)
+ */
+void
+remove_entry_from_bbf_partition_function(char *partition_function_name)
+{
+	Relation	rel;
+	HeapTuple	tuple;
+	ScanKeyData	scanKey[2];
+	SysScanDesc	scan;
+	int16		dbid = get_cur_db_id();
+	int32		partition_function_exists = false;
+	bool		has_dependent_objects = true;
+
+	/* Fetch the relation */
+	rel = table_open(get_bbf_partition_function_oid(), RowExclusiveLock);
+	
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_partition_function_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_partition_function_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(partition_function_name));
+	/* scan using index */
+	scan = systable_beginscan(rel, get_bbf_partition_function_idx_oid(),
+					false, NULL, 2, scanKey);
+	
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{	
+		partition_function_exists = true;
+		/* remove the entry only if there is no dependent partition scheme on it */
+		if (!is_partition_function_used(partition_function_name))
+		{
+			has_dependent_objects = false;
+			CatalogTupleDelete(rel, &tuple->t_self);
+		}
+	}
+
+	systable_endscan(scan);
+	table_close(rel, RowExclusiveLock);
+
+	/* raise error if it doesn't exists in database */
+	if(!partition_function_exists)
+	{
+		ereport(ERROR, 
+			(errcode(ERRCODE_UNDEFINED_OBJECT),
+				errmsg("Cannot drop the partition function '%s', because it does not exist or you do not have permission.", partition_function_name)));
+	}
+
+	/* raise error if there are dependent partition scheme on it */
+	if(has_dependent_objects)
+	{
+		ereport(ERROR, 
+			(errcode(ERRCODE_UNDEFINED_OBJECT),
+				errmsg("Partition function '%s' is being used by one or more partition schemes.", partition_function_name)));
+	}
+}
+
+/*
+ * partition_function_exists
+ * 		checks if provided partition function name exists in database
+ */
+bool
+partition_function_exists(char *partition_function_name)
+{
+	Relation	rel;
+	HeapTuple	tuple;
+	SysScanDesc	scan;
+	ScanKeyData	scanKey[2];
+	int32		partition_function_id = 0;
+	int16		dbid = get_cur_db_id();
+
+	/* open the relation */
+	rel = table_open(get_bbf_partition_function_oid(), AccessShareLock);
+	
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_partition_function_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_partition_function_name,
+				BTEqualStrategyNumber, InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ, CStringGetTextDatum(partition_function_name));
+	
+	/* scan using index */
+	scan = systable_beginscan(rel, get_bbf_partition_function_idx_oid(),
+					false, NULL, 2, scanKey);
+	
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		bool isnull;
+		partition_function_id = DatumGetInt32(heap_getattr(tuple, Anum_bbf_partition_function_id,
```

Comment:
just add safety check that isnull should not be NULL

Reply:
these columns are defined as NOT NULL in the catalog, safety check will be redundant.